processor/src/advice/mod.rs

```diff
@@ -19,38 +19,29 @@ pub use source::AdviceSource;
 // ADVICE PROVIDER
 // ================================================================================================
 
-// TODO elaborate on why this is non-deterministic (seems it is a simple state machine).
-
-// TODO the `advance_clock` seems to be an internal of the VM and shouldn't necessarily be here.
-
-// TODO this will likely suffer a breaking change as we introduce a generic storage for big Merkle
-// sets.
-// The purpose of this clock is to keep the feedback of the provided in sync with the execution
-// trace of the VM, but on the other hand the VM can control the calls/traces itself, without the
-// need for the advice provider (or any other external component) to keep track of it. If we
-// delegate this control to the advice provider - especially if it is a trait, the user might
-// implement it incorrectly, creating undefined behavior on the VM side (that expects the counter
-// to incrementally increase).
-
-/// Common behavior of advice providers for program execution.
+/// Defines behavior of an advice provider.
 ///
-/// An advice provider supplies non-deterministic inputs to the processor.
+/// An advice provider is a component through which the VM processor can interact with the host
+/// environment. The processor can request nondeterministic inputs from the advice provider (i.e.,
```

Comment:
nit: nondeterministic might be a context-heavy term. We might want to elaborate on that. As I see, we come from the STARK context of non-probabilistic proofs, and this as something outside the deterministic domain of STARK; thus, the advice provider is not to be trusted and will not extend the deterministic properties of STARK proofs.

Reply:
Agreed - but I think the right place is probably [this section](https://0xpolygonmiden.github.io/miden-vm/intro/overview.html#nondeterministic-inputs).